commaai/openpilot
# Micro bounty: visiond PC version

A visiond version for linux, tested on ubuntu 16.04.

### Features
**camera_eon_stream.cc:**
- Connect to Eon streaming server (loggerd --stream), decode the hevc  YUV420 frames using libavcodec API.
- Emulate a rear camera,  dispatch YUV420 frames to the visiond_pc.cc processing thread.

**visiond_pc.cc**
- A clone of visiond.cc
- Launch the emulated  camera, run driving model and posnet model, so that all the rear camera vision pipeline can work as normal.
- Removed dependencies on  android ion, visionbuf and visionimg.
- Removed streaming server support.
- Removed front camera and drive monitoring features.

**Other changes**
- Add some missing header inclusions in model.cc, snpemodel.cc and monitor.cc.
- Use CPU mode for x86_64 in snpemodel.cc.
- Fixed a memory leak in model.cc.
Have tested the code on Eon, so that all the changes will not affect the android version of visiond.

### Build Prerequisites
- Install SNPE and set SNPE_ROOT env.
- Build and install zmq & czmq.
- Install libavcodec-dev.

### How to use
**On Eon**
Launch loggerd and visiond
```bash
loggerd --stream
visiond
```

**On ubuntu**
Launch visiond, set the target Eon IP address using  env EON_IP:
```bash
EON_IP=192.168.1.105 ./visiond
```
Launch calibration daemon to handle posnet inference result and do live calibration:
```bash
python calibrationd.py
```
Launch openpilot-tools replay UI to see the decode frames and model inference result:
```bash
python openpilot-tools/replay/.py
```
### How to verify the model inference result
- Play a recorded drive video on PC, like this:
https://github.com/commaai/comma2k19/blob/master/Example_1/b0c9d2329ad1606b%7C2018-08-02--08-34-47/40/video.hevc
- Point the Eon at the screen, make sure the lanes is within the blue box on openpilot-tools replay UI.
- Check the model output window, will see that the lanes is detected correctly by driving model.
![](https://i.ibb.co/LZZ4ScH/eon-streaming.jpg)
